src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Setup/SetupComponentRegistrationService.cs

```diff
@@ -0,0 +1,259 @@
+
+            string[]? runtimeVersions = NetCoreRuntimeVersionsRegistryReader.ReadRuntimeVersionsInstalledInLocalMachine(architecture);
+
+            if (runtimeVersions is null)
+            {
+                return null;
+            }
+
+            HashSet<string> installedRuntimeComponentIds = new(StringComparers.VisualStudioSetupComponentIds);
+
+            foreach (string runtimeVersion in runtimeVersions)
+            {
+                // Example versions:
+                //
+                // - "3.1.32"
+                // - "7.0.11"
+                // - "8.0.0-preview.7.23375.6"
+                // - "8.0.0-rc.1.23419.4"
+                //
+                // We only want the major/minor version, in the same format as TargetFrameworkVersion (e.g. "v8.0").
+
+                int firstDotIndex = runtimeVersion.IndexOf('.');
+                int secondDotIndex = runtimeVersion.IndexOf('.', firstDotIndex + 1);
+                string versionNumber = runtimeVersion.Substring(0, secondDotIndex);
+                string version = $"v{versionNumber}";
+
+                if (ConfiguredSetupComponentSnapshot.ComponentIdByRuntimeVersion.TryGetValue(version, out string? runtimeComponentId))
+                {
+                    installedRuntimeComponentIds.Add(runtimeComponentId);
+                }
+            }
+
+            return installedRuntimeComponentIds;
+        }
+    }
+
+    protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+    {
+        _solutionEventsSubscription = await _solutionService.SubscribeAsync(this, cancellationToken);
+
+        _setupCompositionService = await _setupCompositionServiceFactory.GetValueAsync(cancellationToken);
+
+        IBrokeredServiceContainer serviceBrokerContainer = await _serviceBrokerContainer.GetValueAsync(cancellationToken);
+        IServiceBroker serviceBroker = serviceBrokerContainer.GetFullAccessServiceBroker();
+        _missingComponentRegistrationService = await serviceBroker.GetProxyAsync<IMissingComponentRegistrationService>(
+            serviceDescriptor: VisualStudioServices.VS2022.MissingComponentRegistrationService,
+            cancellationToken: cancellationToken);
+    }
+
+    protected override async Task DisposeCoreAsync(bool initialized)
+    {
+        if (_solutionEventsSubscription is not null)
+        {
+            await _solutionEventsSubscription.DisposeAsync();
+        }
+
+        if (_missingComponentRegistrationService is IDisposable disposable)
+        {
+            disposable.Dispose();
+        }
+    }
+
+    public async Task<IDisposable> RegisterProjectAsync(Guid projectGuid, CancellationToken cancellationToken)
+    {
+        Requires.Argument(projectGuid != Guid.Empty, nameof(projectGuid), "Cannot be an empty GUID.");
+
+        await InitializeAsync(cancellationToken);
+
+        // Add an empty snapshot for the project for now.
```

Comment:
Why is this done?

Reply:
The component calls VS Setup, passing data from all projects. We try to avoid that call until we have data from all projects. So when a project registers itself we track it with the null. Then later, in `HaveAllSnapshots`, we ensure there are no null values before publishing.

I'll include this information in a comment on the collection field.